### PR TITLE
Align plugin/template usage and syntax with other plugins

### PR DESCRIPTION
**Pull request by rtreffer**
### 1. What does this pull request do?

The plugin/template is currently scoped by regex with a fallthrough default.

This patch changes the config to be zone scoped and makes fallthrough an option instead of a default.

### 2. Which issues (if any) are related?

This change was discussed in #1298

### 3. Which documentation changes (if any) need to be made?

README change is included, plugin/template wasn't shipped yet.

### CODE REVIEW

**Comment by rtreffer** [comment]
@miekg hot off the press. This is an alternative syntax scoped by zone with an explicit regex match.

File: plugin/template/README.md [hunk 1]
```diff
@@ -9,29 +9,34 @@ The *template* plugin allows you to dynamically repond to queries by just writin
...
 
-* **CLASS** the query class (usually IN or ANY)
-* **TYPE** the query type (A, PTR, ...)
+* **ZONE** the zone scope(s) for this template. Defaults to '.' (everything).
+* **CLASS** the query class (usually IN or ANY).
+* **TYPE** the query type (A, PTR, ... can be ANY to match all types).
 * **REGEX** [Go regexp](https://golang.org/pkg/regexp/) that are matched against the incoming question name. Specifying no regex matches everything (default: `.*`). First matching regex wins.
 * `RR` A [RFC 1035](https://tools.ietf.org/html/rfc1035#section-5) style `<rr>` fragment build by a [Go template](https://golang.org/pkg/text/template/) that contains the answer.
 * `responsecode` A response code (`NXDOMAIN, SERVFAIL, ...`). The default is `SUCCESS`.
+* `fallthrough` Continue with the next plugin if the zone matched but no regex did not match.
```
Comment (miekg) [inline plugin/template/README.md]: This also has a optional ZONES argument as well (as of yesterday evening) :)

File: plugin/template/README.md [hunk 2]
```diff
@@ -9,29 +9,34 @@ The *template* plugin allows you to dynamically repond to queries by just writin
...
     [authority RR]
     [...]
     [rcode responsecode]
+    [fallthrough]
 }
 ~~~
 
-* **CLASS** the query class (usually IN or ANY)
-* **TYPE** the query type (A, PTR, ...)
+* **ZONE** the zone scope(s) for this template. Defaults to '.' (everything).
```
Comment (miekg) [inline plugin/template/README.md]: Put zone below **TYPE** as it is last here

File: plugin/template/README.md [hunk 3]
```diff
@@ -9,29 +9,34 @@ The *template* plugin allows you to dynamically repond to queries by just writin
...
+* **CLASS** the query class (usually IN or ANY).
+* **TYPE** the query type (A, PTR, ... can be ANY to match all types).
 * **REGEX** [Go regexp](https://golang.org/pkg/regexp/) that are matched against the incoming question name. Specifying no regex matches everything (default: `.*`). First matching regex wins.
```
Comment (miekg) [inline plugin/template/README.md]: `match` **REGEX**

File: plugin/template/README.md [hunk 4]
```diff
@@ -63,7 +84,7 @@ The `.invalid` domain is a reserved TLD (see [RFC-2606 Reserved Top Level DNS Na
 . {
     proxy . 8.8.8.8
 
-    template ANY ANY "[.]invalid[.]$" {
+    template ANY ANY invalid. {
```
Comment (miekg) [inline plugin/template/README.md]: Should we make this final dot implicit?

File: plugin/template/template_test.go [hunk 5]
```diff
@@ -260,8 +226,12 @@ func TestHandler(t *testing.T) {
 
 	for _, tr := range tests {
 		handler := Handler{
-			Next:      test.NextHandler(rcodeFallthrough, nil),
-			Templates: []template{tr.tmpl},
+			Next:        test.NextHandler(rcodeFallthrough, nil),
+			Class:       tr.qclass,
+			Qtype:       tr.qtype,
+			Zones:       []string{"."},
+			Fallthrough: true,
```
Comment (miekg) [inline plugin/template/template_test.go]: See the new `fall` pkg in plugin/pkg/fall for the new way of doing fallthrough

**Comment by miekg** [comment]
Let me merge #1361, but this PR looks good.

File: plugin/template/README.md (hunk 4 above)
Comment (rtreffer) [reply plugin/template/README.md]: Good question, I'm using the same logic as etcd.... And I'm usually trying to mark qualified names as such.

Related: If nothing is specified I'm right now matching '.' - is there a reason to switch to the etcd/setup.go method of
```
		if len(etc.Zones) == 0 {
			etc.Zones = make([]string, len(c.ServerBlockKeys))
			copy(etc.Zones, c.ServerBlockKeys)
		}
```

And it looks like I should normalize, too (according to the etcd source)

**Comment by rtreffer** [comment]
@miekg rebased on master and most points fixed.

Metrics are broken, I'll need another PR for that (and #1362 - remove regex from metric labels)

File: plugin/template/README.md (hunk 4 above)
Comment (miekg) [reply plugin/template/README.md]: yes, this should be `Normalized`

File: plugin/template/README.md [hunk 6]
```diff
@@ -88,14 +110,33 @@ path (`dc1.example.com`) added.
...
+    template IN ANY example.com. {
+      match "(example.com.dc1.example.com)$"
       rcode NXDOMAIN
       answer "{{ index .Match 1 }} 60 IN SOA a.{{ index .Match 1 }} b.{{ index .Match 1 }} (1 60 60 60 60)"
+      fallthrough
     }
 }
 ~~~
 
+<<<<<<< HEAD
```
Comment (miekg) [inline plugin/template/README.md]: This didn't survive the rebase.

File: plugin/template/README.md [hunk 7]
```diff
@@ -9,29 +9,34 @@ The *template* plugin allows you to dynamically repond to queries by just writin
...
-template CLASS TYPE [REGEX...] {
+template CLASS TYPE [ZONE...] {
+    [match REGEX...]
     [answer RR]
     [additional RR]
     [authority RR]
     [...]
-    [rcode CODE]
+    [rcode responsecode]
+    [fallthrough]
```
Comment (miekg) [inline plugin/template/README.md]: fallthrough can now have optional ZONES

File: plugin/template/README.md (hunk 4 above)
Comment (miekg) [reply plugin/template/README.md]: drop the final dot (as this is not needed anymore)

File: plugin/template/setup_test.go [hunk 8]
```diff
@@ -93,30 +93,37 @@ func TestSetupParse(t *testing.T) {
...
 			}`,
 			false,
 		},
 		{
-			`template IN ANY "[.](example[.]com[.]dc1[.]example[.]com[.])$" {
+			`template IN ANY example.com. {
+				match "[.](example[.]com[.]dc1[.]example[.]com[.])$"
 				rcode NXDOMAIN
 				answer "{{ index .Match 1 }} 60 IN SOA a.{{ index .Match 1 }} b.{{ index .Match 1 }} (1 60 60 60 60)"
+				fallthrough example.com.
```
Comment (miekg) [inline plugin/template/setup_test.go]: doesn't need closing dot

File: plugin/template/template.go [hunk 9]
```diff
@@ -108,21 +132,16 @@ func executeRRTemplate(section string, template *gotmpl.Template, data templateD
 	return rr, nil
 }
 
-func (t template) match(state request.Request) (templateData, bool) {
+func (t template) match(state request.Request, zone string) (templateData, bool) {
 	q := state.Req.Question[0]
 	data := templateData{}
 
-	if t.class != dns.ClassANY && q.Qclass != dns.ClassANY && q.Qclass != t.class {
```
Comment (miekg) [inline plugin/template/template.go]: We don't need this shortcut any more?

**Review by miekg — CHANGES_REQUESTED** [review]

**Comment by miekg** [comment]
btw: I can wait with 1.0.3 until this is sorted out; there is not actual rush for 1.0.3 other than getting the code out there. Having a new plugin in a release always looks good :)

**Comment by rtreffer** [comment]
Just got up. There is an issue in this PR around mutliple sections scoped by different zones....

I'll change this the following way: the plugin does a very rough / quick check (can I be responsible for this zone at all?) and the template does a more thorough one (actual zone, type, regex).

I'm adding a test for that now.

**Comment by miekg** [comment]
> I'll change this the following way: the plugin does a very rough / quick check (can I be responsible for this zone at all?) and the template does a more thorough one (actual zone, type, regex).

This is how other plugins do this as well, so lgtm

**Comment by rtreffer** [comment]
I'm now more carefully scoping the match checks and added a test to verify it works: https://github.com/soundcloud/coredns/blob/f047cb0085c3935e7491adb7541d2929189af557/plugin/template/template_test.go#L300

I still have to do the metrics change and the mentioned README cleanup.

**Comment by rtreffer** [comment]
@miekg everything should be addressed, including #1360

**Review by miekg — APPROVED** [review]

**Pull request merged**